Fix #2446: latex(pdf) sets local tables of contents in unbreakable boxes

jfbu opened the pull request:
Like framed.sty is used for framed literal-blocks allowing page breaks, this uses in another way a custom framed environment to allow page breaks in tables of contents generated by `contents` directive with `local` option.

The former use of fancybox.sty's macro `\shadowbox` was buggy from the LaTeX point of view, and extended in an uncontrolled way into the right margin. The proposal puts the (by default) 4pt width shadow in the margin with the main frame exactly at the right text margin.

To remove all dependencies on fancybox.sty we also replace used of its Sbox environment for Notices (warnings, etc..) with a framed environment. Currently the minipage is maintained hence this can't break across pages but could be added.

The PR introduces slight variation of vertical spacing (a bit more space around admonitions, a little less after tables of contents boxes).

Fixes #2446. This commit is based upon earlier PR #2448.

jfbu committed:
Fix #2446: latex(pdf) sets local tables of contents in unbreakable boxes

Allow page breaks in topic contents (e.g. local tables of contents) from
replacing \shadowbox of non anymore used package framebox by a custom
framed package environment.

Formerly, topic boxes extended into right margin; put only the shadow in
margin, locate the right frame exactly at the margin.

Use framed.sty also for framing admonitions of the warning type. Make
vertical spacing more coherent with the one for hint type notices.
Handle case with warning notices getting nested; the inner ones use a
minipage, only the outer one allows pagebreaks.

jfbu commented:
@tk0miya I had a bit of trouble of obtaining correct way to mimic earlier behaviour via `center` environment and a `\shadowbox` wider than textwidth. At last I think I now have the correct way to have almost same vertical spaces. As per horizontal, I think that the fact that `topic` boxes extended into right margin was an unintended feature. I find it nice looking if only the lateral shadow of the box actually falls into margin. But it is also possible to recover former looks if needed.

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -282,6 +255,70 @@
+   \def\FrameHeightAdjust {\baselineskip}%
+   % the \noindent is needed to get proper vertical spacing at the
+   % \endtrivlist location after \MakeFramed. Together with \FrameHeightAdjust
+   % setting, exact same vertical spacing as with Sphinx 1.4.1 is obtained.
```
well not "exact" but close enough

jfbu commented:
@tk0miya I rebased and squashed commits in this branch to have only one commit. Sorry for inconvenience ;-)

jfbu commented:
well, I needed a second commit to fix the broken behaviour in parent commit of `warning` and similar directives in indented contexts. On this occasion I now remove from `\py@heavybox` use of minipage encapsulation. Thus, not only `hint` types but also `warning` types of admonitions may, if there are long be split on a page break. Also, behaviours with regards to vertical space after admonition notice is unified.

jfbu commented:
I have squashed all into one single commit (the material related to `\py@heavybox` was initially wrong, I decided it was better to not leave a buggy commit, and merge the fix into it).

tk0miya commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -11,14 +11,17 @@
 \RequirePackage{textcomp}
-\RequirePackage{fancybox}
+% fancybox was used in Sphinx <= 1.4.1 for \shadowbox and Sbox environment.
+% \RequirePackage{fancybox}
```
To keep dependency carefully, could you keep loading the `fancybox` package? And give a comment as "fancybox will be removed at Sphinx-1.5" please.
It might break the documents if it use the package in their customization and extension. We should keep compatibility during same major version. Probably nobody depends it :-p

After that, please remove the `RequirePackage` definition in master branch with append that to "incompatible changes" section on CHANGES file.

tk0miya commented:
Could you add a comment to CHANGES file?

tk0miya commented:
Note (this does not prevent to merge this PR):
I think keeping box shadowed is not important. We can move other package if there is any good substitution package which can represent topic block (of course, it must be in master branch).
It would be nice if it will be replaced external package. Because LaTeX macro is too difficult to me to maintain them.

jfbu committed:
Merge branch 'stable' into fixpdfunbreakabletopicboxes

jfbu committed:
Update CHANGES for PR #2453

jfbu commented:
@tk0miya  advantage of `framed.sty` package is that it has been available for a long time and has no dependencies and is small package; but its documentation is sparse, and its customization reserved to advanced LaTeX users. 

Advanced box customization is available via package `tcolorbox` (http://ctan.org/pkg/tcolorbox) but it loads `pgf` (underlying graphics layer of `TikZ`) which is big hence slows down significantly LaTeX run. Besides, all the customization is not really useful if no interface is provided to user; and then there is the question that the options should work or make sense in all builders. Ultimately Sphinx will probably switch to these more modern tools I guess.

I have squashed the commits in this PR. I have added comments to explain more the code.

From the LaTeX point of view it is now possible to nest topics ; but this is disallowed by LaTeX builder. Nesting of Warnings does work. The innermost ones will be typeset in an unbreakable minipage, else compilation fails; only the outermost one allows pagebreaks.

I have tested behaviours in Quote-like and Lists environment and all seems ok. What do you think about merge?

jfbu committed:
fix a latex issue of vertical space

	modified:   sphinx/texinputs/sphinx.sty

tk0miya commented:
Thank you for comment. I understand why we use `framed.sty`.

Now it looks good to me. Merge this please :-)

jfbu commented:
@tk0miya done :-) but I merged from the web interface hence I see my email is anonymized .. 

about `framed` I agree with you it requires very advanced LaTeX understanding (at the level of deeply understanding how LaTeX does lists for example). As I worked on it these last few days I slowly understood some subtleties of its interface. This package will not change, and if it changes it will certainly not break things. With some more recent packages relying on TikZ this is less obvious. They require much more resources and once in a while there are incompatible changes. Nevertheless at some point in the future, it will probably appear useful to switch, as TikZ now is quite mature.

I could at some point in the future create a package for extensions to framed.sty with the things like shadowed boxes we need in sphinx.sty, and contribute it to the package repository for TeX (http://ctan.org/). Then the `sphinx.sty` would become leaner ... but people would need a recent TeX install ... It is much easier nowadays than ten years ago to have an up-to-date TeX install, but still some Linux distros often have very outdated (like 4 or 5 years) packages.

jfbu commented:
@tk0miya sorry to ping you again but I am learning my trade here ;-) should I merge stable into master to then remove there the `\RequirePackage{fancybox}` and update the CHANGES about it ?

tk0miya commented:
Yes, please. You can merge stable into master when always you need.

BTW, sphinx now bundles some LaTeX packages. So we can bundle your package if you will release it :-)

tk0miya commented:
I don't worry about `framed.sty` seriously.
In addition, we would have good support from you :-)

Thank you for explaining it so clearly!
